### ISSUE TITLE
Firefox audio recording quality is unusably bad

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#8415]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### Qubes OS release

R4.2

### Brief summary

The audio quality of Zoom meetings is unusably bad.

### Steps to reproduce

Join a Zoom meeting.  I’m using PipeWire, in case that matters.

### Expected behavior

Audio quality is good.

### Actual behavior

Audio quality is unusable.

**Comment by DemiMarie** [comment]
Switching from PipeWire to legacy PulseAudio fixes the problem, so this is a PipeWire agent problem.

**Comment by ZYousef** [comment]
I'm not sure if this is relevant but I've noticed improved (https://github.com/QubesOS/qubes-issues/issues/8001) microphone quality during calls with Pipewire agents.

**Comment by DemiMarie** [comment]
This turned out to be a problem with the Zoom Flatpak.  The official Zoom RPMs are fine, so long as Zoom has enough vCPUs for software rendering to keep up.

Closing as “not our bug” (the Qubes OS Project does not produce the software responsible for this bug).

**Comment by Minimalist73** [comment]
@DemiMarie I have the same problem on 4.2, but it is system wide. The voice is not clear, cuts out from time to time, making it impossible for the other people in the call to hear me properly. Depending on the application (Firefox / Signal) or system (debian or fedora), I experience some latency between when I speak and when the application gets it. I switched back to pulseaudio like you did and it fixed all these issues.
I use a jack headset (mic and audio input separated) and I did not change any option in dom0 related to audio. Any ideas?

**Comment by DemiMarie** [comment]
@Minimalist73 What is the output of `ps -f $(pgrep pacat)`?

**Comment by Minimalist73** [comment]
@DemiMarie 

Here is the command result on my test qube running pipewire on fedora-38-xfce:

```
[user@test-fedora ~]$ ps -f $(pgrep pacat)
UID          PID    PPID  C STIME TTY          TIME CMD
user        1536     965  0 20:32 pts/1    00:00:00 bash
user        1662    1536  0 20:34 pts/1    00:00:00 ps -f
```

Same for my debian-12-xfce test qube.

**Comment by Minimalist73** [comment]
@DemiMarie Sorry, here is the result when I run the command in dom0:

```
[User@dom0 ~]$ ps -f $(pgrep pacat)
UID          PID    PPID  C STIME TTY      STAT   TIME CMD
User        4350    4098  0 13:06 ?        Sl     0:00 /usr/bin/pacat-simple-vchan -l 3 sys-usb
User        6330    4098  0 13:09 ?        Sl     0:00 /usr/bin/pacat-simple-vchan -l 8 sys-net
User        6381    4098  0 13:09 ?        Sl     0:00 /usr/bin/pacat-simple-vchan -l 10 sys-firewall
User       35775    4098  0 14:58 ?        Sl     0:00 /usr/bin/pacat-simple-vchan -l 24 sys-whonix
User      142424    4098  0 20:50 ?        S      0:00 /usr/bin/pacat-simple-vchan 60 test-debian
User      143453    4098  2 20:51 ?        S      0:11 /usr/bin/pacat-simple-vchan 61 test-fedora
```

I restarted both qubes recently, only retested on fedora for now and I still have the issue right now.

**Comment by DemiMarie** [comment]
@Minimalist73 It looks like you have <https://github.com/QubesOS/qubes-core-admin-client/pull/261>, but that clearly is not solving the problem.

I changed the title of this issue to be about what I suspect the real solution is: the PipeWire agent needs to adapt to overruns and underruns by increasing the audio graph quantum, thereby allowing `pacat-simple-vchan` in the AudioVM to catch up.  It also needs to lower the quantum when things are good so that the user isn’t stuck with high latency forever.  This is going to require some fairly complex logic, and PipeWire doesn’t (yet) have any useful utility functions to help with it.

**Comment by Minimalist73** [comment]
@DemiMarie Since it's not yet possible to do something about it, I'll switch back my templates to pulseaudio in the meantime. Thank you for taking the time to look into my problem.

**Comment by marmarek** [comment]
This turned out to be very bad, affecting most of Firefox when using with PipeWire. Marking it as a release blocker. We'll try to fix the pipewire agent, but if that won't work quickly enough, we'll switch back to pulseaudio (which is a bit painful because requires manual step).

**Comment by marmarek** [comment]
FWIW PipeWire agent with Chromium works just fine. The problematic case is audio recording (microphone for audio calls) with Firefox.

**Comment by marmarek** [comment]
Interestingly, I do not see many "Underrun" messages - I see some, but about 2 or 3 per minute, while the audio breaks happen a lot - in fact it's mostly no sound at all, and only some audio works every few seconds.

And another thing tested - downgrading to pipewire 0.3.67 (the other one available in fedora repository) did not helped.

**Comment by DemiMarie** [comment]
What about Firefox ESR on Debian?

**Comment by Minimalist73** [comment]
I did a few test with Firefox, Firefox ESR and Chromium using the edit made in the pull request QubesOS/qubes-gui-agent-linux/pull/199 and I have a strange effect:
- When I attach my microphone to the qube while the browser is open, the voice cuts and sometime can't be heard.
- When I attach my microphone before I start the browser, my voice still cuts but after a few seconds it comes clear and works perfectly.

This happens on all browsers I tested and I replicated it multiple times. 

While doing my tests, pipewire was killed a few times while I was attaching/detaching my microphone (qube is running Debian 12 with 4 vCPU and 4GB RAM):
```
test-debian pipewire[641]: free(): invalid next size (normal)
test-debian systemd[616]: pipewire.service: Main process exited, code=killed, status=6/ABRT
test-debian systemd[616]: pipewire.service: Failed with result 'signal'.
test-debian systemd[616]: pipewire.service: Consumed 25.665s CPU time.
test-debian systemd[616]: pipewire.service: Scheduled restart job, restart counter is at 1.
test-debian systemd[616]: Stopped pipewire.service - PipeWire Multimedia Service.
test-debian systemd[616]: pipewire.service: Consumed 25.665s CPU time.
test-debian systemd[616]: Started pipewire.service - PipeWire Multimedia Service.
test-debian systemd[616]: Stopping pipewire.service - PipeWire Multimedia Service...
test-debian pipewire[3687]: Asked to handle disabled watch: 0x5c2f2370d870 19
test-debian systemd[616]: Stopped pipewire.service - PipeWire Multimedia Service.
```

**Comment by DemiMarie** [comment]
@Minimalist73 Please file a separate issue.  This particular issue is a bug in either Firefox, pipewire-pulse, or both.

**Comment by andrewdavidwong** [comment]
Since this is not a bug in Qubes code (and therefore out of our control to fix), let's lower the priority on this issue from "blocker" to "critical" so as not to block the 4.2 release.

**Comment by GammaSQ** [comment]
How does one "switch to PulseAudio"?

This is still a big issue for me. Brave (chromium based) has this issue. Attaching microphone before starting browser did not fix the issue. dom0 is fully updated, sys-audio is based on fedora-38-minimal.

**Comment by DemiMarie** [comment]
@GammaSQ Install `pulseaudio-qubes` in the TemplateVM.

**Comment by bickelj** [comment]
My apologies for necroing here, let me know if I should put this elsewhere. I found the above post actionable so I put my experience with it below it.

In the case of Qubes 4.3 with Debian 13 trixie standalone VM, when I `apt install pulseaudio-qubes` it uninstalls `qubes-vm-recommended`:

```
Installing:
  pulseaudio-qubes

REMOVING:
  pipewire-qubes  qubes-vm-recommended

```

Then a litany of good stuff will autoremove:

```
$ sudo apt autoremove
REMOVING:                       
  debugedit       libmodulemd2       python3-aiohappyeyeballs  python3-frozenlist          python3-numpy              python3-rpm          qubes-core-agent-dom0-updates       rpm2cpio
  deltarpm        libraqm0           python3-aiohttp           python3-gnupg               python3-numpy-dev          python3-simplejson   qubes-core-agent-passwordless-root  salt-common
  dnf             librepo0           python3-aiosignal         python3-gpg                 python3-olefile            python3-tempora      qubes-gpg-split                     salt-ssh
  dnf-data        librpm-sequoia-1   python3-apt               python3-hawkey              python3-pil                python3-tornado      qubes-img-converter                 sqlite3
  fwupd-qubes-vm  librpm10           python3-async-timeout     python3-jaraco.collections  python3-portend            python3-tqdm         qubes-input-proxy-sender            usbutils
  jcat            librpmbuild10      python3-attr              python3-jaraco.text         python3-propcache          python3-tz           qubes-mgmt-salt-vm-connector
  libcomps0       librpmio10         python3-cheroot           python3-legacy-cgi          python3-pycares            python3-unbound      qubes-notification-agent
  libdnf2-common  librpmsign10       python3-cherrypy3         python3-libcomps            python3-pycryptodome       python3-virtualenv   qubes-pdf-converter
  libdnf2t64      libsolv1           python3-croniter          python3-libdnf              python3-pytz               python3-webob        qubes-repo-templates
  libfsverity0    libsolvext1        python3-distlib           python3-looseversion        python3-qubesimgconverter  python3-yarl         qubes-usb-proxy
  libjcat1        python-apt-common  python3-dnf               python3-msgpack             python3-repoze.lru         python3-zc.lockfile  rpm
  liblua5.3-0     python3-aiodns     python3-filelock          python3-multidict           python3-routes             python3-zmq          rpm-common

```

To work around this, I manually installed the ones with `qubes` in the name to prevent auto-removal:

`sudo apt install qubes-core-agent-dom0-updates qubes-core-agent-passwordless-root qubes-gpg-split qubes-img-converter qubes-input-proxy-sender qubes-mgmt-salt-vm-connector qubes-notification-agent qubes-pdf-converter qubes-repo-templates qubes-usb-proxy fwupd-qubes-vm`

Then all is well:

```bash
$ sudo apt autoremove
Summary:                        
  Upgrading: 0, Installing: 0, Removing: 0, Not Upgrading: 0
```

I hope this helps someone who notices the same.

I also see that in Debian 13 trixie pipewire is the default so installing `pulseaudio-qubes` and removing `pipewire-qubes` may not be sufficient to "use pulseaudio" because it leaves/uses pipewire's pulse adapter (`pipewire-pulse`). Furthermore, doing this solely in the non-audio VM wouldn't affect the audio VM. Take all this with a huge grain of salt.

(TBD: whether this helps audio quality in Firefox)

**Comment by unman** [comment]
This is useful in *any* case where you want to remove something part of,
or a dependency on, qubes-vm-recommended.

In this case, though, trixie should use pipewire-qubes - why did you
find it necessary to install pulseaudio-qubes?

**Comment by unman** [comment]
Ah, I've just picked up on the subject line. Did installing pulseadio improve
the sound quality for you?
